Renew build guide (#725)
PowerGridModel/power-grid-model: docs/advanced_documentation/build-guide.md

* A `power-grid-model` [pip](https://pypi.org/project/power-grid-model/) Python package with C++ extension as the calculation core.
* A [CMake](https://cmake.org/) project consisting of the C++ header-only calculation core, and the following build targets:
    * A dynamic library (`.dll` or `.so`) with stable pure C API/ABI which can be used by any application (enabled by default)

>     * An install target that installs a package that contains the dynamic library (enabled by default)

Reviewer: ```suggestion
    * An install target that installs a package containing the dynamic library (enabled by default)
```

> you can install the library from source in develop mode with the development dependency. 

Reviewer: ```suggestion
you can install the library from source in development mode with the development dependency. 
```
I think the official term for this is either "development mode" or "editable installs"

Go to the root folder of the repository.

```shell
pip install -e .[dev]
```

Then you can run the tests.

```shell
pytest
```

A basic `self_test` function is provided to check if the installation was successful and ensures there are no build errors, segmentation violations, undefined symbols, etc. It performs multiple C API calls, runs through the main data flow, and verifies the integrity of serialization and deserialization.

```python
from power_grid_model.utils import self_test
self_test()
```

## Build CMake Project

### User build

If you are a C-API user of the library, you can build the CMake using all the default settings.
You can specifiy a standard [CMAKE_BUILD_TYPE](https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html).
This will only build the core C-API dynamic library.

```shell
cmake -DCMAKE_BUILD_TYPE=Release -B build
cmake --build build --config Release
```

### Developer build

> If you are a developer of Power Grid Model, 

Reviewer: ```suggestion
If you opt for a developer build of Power Grid Model, 
```
This may not necessarily be true? A developer of PGM sounds like officially affiliated titles ;)

you can use the pre-defined CMake presets to enable developer build, including all the tests, warnings, examples, and benchmark. In the presets the [Ninja](https://ninja-build.org/) generator is used. 
In principle, you can use any C++ IDE with cmake and ninja support to develop the C++ project.
It is also possible to use the bare CMake CLI to set up the project. 
Supported presets for your development platform can be listed using `cmake --list-presets`.

> In the developer build the following build targets (folders) are enabled:

Reviewer: ```suggestion
In the developer build the following build targets (directories) are enabled:
```

* `power_grid_model_c`: a dynamic library (`.dll` or `.so`) with stable pure C API/ABI which can be used by any application
* `tests/cpp_unit_tests`: the unit test target for the C++ core using the `doctest` framework.
* `tests/cpp_integration_tests`: the integration test target for the C++ core using the `doctest` framework.
* `tests/cpp_validation_tests`: the validation test target using the `doctest` framework

> * `tests/native_api_tests`: the C API test target using the `doctest` framework

Reviewer: Oops..
